tools/fp20compiler/main.cpp

```diff
@@ -66,6 +70,20 @@ void translate(const char* s) {
         char* line_end = strchr(shader_start, '\n');
         char* next_shader_start = find_at_line_start(ns, line_end + 1, "!!");
 
+        // Create a copy of the current shader magic
+        if (line_end) { *line_end = '\0'; }
+        char* shader_magic = strdup(shader_start);
+        if (line_end) { *line_end = '\n'; }
+
+        // Count which line the current shader starts on
+        while (line_cursor < shader_start) {
+            line_cursor = strchr(line_cursor + 1, '\n');
+            line++;
+        }
```

Comment:
Is `line` meant to specify the line number of the magic or the start of the actual code? It currently does the latter.

Reply:
The shader magic is part of the shader code already.

`line` is meant to be the line which contains the shader magic.
I believe this worked fine for me?

Reply:
It seems to be off by one - when testing with the triangle sample, I get [this cgc output](https://pastebin.com/3QrbdtRW), and [this](https://pastebin.com/6qFjpPgi) is the end result, which lists the number of the first line after the magic.

Reply:
There was an off-by-one. I had to change the algorithm slightly and added an `assert` for safety.